### ISSUE TITLE
Unable to import Google Style Gude ruleset

### ISSUE THREAD

**Issue by romani** [checkstyle/sonar-checkstyle#5]
https://github.com/SonarQubeCommunity/sonar-checkstyle/issues/31

 jakub-bochenski commented 12 days ago

When trying to import https://raw.githubusercontent.com/checkstyle/checkstyle/master/src/main/resources/google_checks.xml I get

>    We're sorry, but something went wrong.
>     Please try back in a few minutes and contact support if the problem persists.
>     Go back to the homepage

@jakub-bochenski
jakub-bochenski commented 12 days ago

I think this is the cause:

`2016.04.14 12:39:59 ERROR web[o.s.s.ui.JRubyFacade] Fail to render: https://sonar.egnyte-internal.com/profiles/create
The definition of the profile 'null' (language 'null') contains multiple occurrences of the 'checkstyle:com.puppycrawl.tools.checkstyle.checks.blocks.RightCurlyCheck' rule. The plugin which declares this profile should fix this.`

@jakub-bochenski
jakub-bochenski commented 12 days ago

I had to remove

```
        <module name="AnnotationLocation">
            <property name="tokens" value="VARIABLE_DEF"/>
            <property name="allowSamelineMultipleAnnotations" value="true"/>
        </module>
```

and

```
        <module name="RightCurly">
            <property name="option" value="alone"/>
            <property name="tokens" value="CLASS_DEF, METHOD_DEF, CTOR_DEF, LITERAL_FOR, LITERAL_WHILE, LITERAL_DO, STATIC_INIT, INSTANCE_INIT"/>
        </module>
```

Then it failed to create the profile with this error:

`Rule template can't be activated on a Quality profile: checkstyle:com.puppycrawl.tools.checkstyle.checks.coding.IllegalTokenTextCheck
127.0.0.1 - - [14/Apr/2016:13:15:48 +0000] "POST /profiles/create HTTP/1.1" 302 108 "https://sonar.egnyte-internal.com/profiles" "Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:43.0) Gecko/20100101 Firefox/43.0"
`

@jakub-bochenski
jakub-bochenski commented 12 days ago

Removing this:

```
        <module name="IllegalTokenText">
            <property name="tokens" value="STRING_LITERAL, CHAR_LITERAL"/>
            <property name="format" value="\\u00(08|09|0(a|A)|0(c|C)|0(d|D)|22|27|5(C|c))|\\(0(10|11|12|14|15|42|47)|134)"/>
            <property name="message" value="Avoid using corresponding octal or Unicode escape."/>
        </module>
```

and

```
        <module name="NoLineWrap"/>
```

Results in

`Value ' LITERAL_FINALLY' must be one of : LITERAL_WHILE, LITERAL_TRY, LITERAL_CATCH, LITERAL_FINALLY, LITERAL_DO, LITERAL_IF, LITERAL_ELSE, LITERAL_FOR, INSTANCE_INIT, STATIC_INIT, LITERAL_SWITCH, LITERAL_SYNCHRONIZED, LITERAL_CASE, LITERAL_DEFAULT, ARRAY_INIT.
`

Which doesn't make sense.

@jakub-bochenski
jakub-bochenski commented 12 days ago

Removing spaces between elements fixed that resulting in:

>    Value ' TEXT' must be one of : text,stmnt

after lowercasing it I hit

`Rule template can't be activated on a Quality profile: checkstyle:com.puppycrawl.tools.checkstyle.checks.blocks.NeedBracesCheck
`

so I've removed

```
        <module name="NeedBraces"/>
```

resulting in:
`
    Rule template can't be activated on a Quality profile: checkstyle:com.puppycrawl.tools.checkstyle.checks.blocks.LeftCurlyCheck
`

I'm giving up now as the ruleset is useless with so many changes.

**Comment by timwebster9** [comment]
Hi,

I've run into this as well.  Can I ask if this is something that is planned to be worked on?

It seems like this would be VERY useful.  This style guide is something that I imagine a lot of companies will adopt (if they haven't already).  I am considering installing the Checkstyle plugin just to do this for our company.

If the SonarQube quality profile import is always going to be a problem, does anyone know if it's possible to just set up the Google rules manually?

**Comment by jakub-bochenski** [comment]
@timwebster9 I've posted a gist with Maven's pom.xml that you can use to enforce the rules via checkstyle. It takes seconds to run.

https://gist.github.com/jakub-bochenski/3e3b2acd0c01ecd884e2d0e4413cb56f

We run it on gitlab-ci to ensure every branch has proper code -- something that'd be very hard to do with sonar (time taken, general crappy branch handling).
There is an additional profile that will make it more verbose (activates by default on Jenkins)

The above also works great with Eclipse m2e and checkstyle plugins -- no need to manually set anything, configuration is automatically kept in sync with pom.xml

**Comment by timwebster9** [comment]
Hi Jakub,

Thanks for the suggestion..:-)  Can I ask some questions:

- I'm guessing that since it fails the build, what process do you then follow?  Does the developer then look at the failed build and see the violations in the build log?  Is it easy to tell what the problem is?
- Have you tried to set up the Google style rules manually in SonarQube?  I know it sounds like a pain, but there aren't that many rules and was just wondering if it's possible...

Cheers,

Tim

**Comment by jakub-bochenski** [comment]
Gitlab-CI let's you run builds for each Pull Request. So yes, the developer will be notified about the failed build and furthermore it will not be possible to merge it if the build is not OK.

Sample output:
```
[INFO] There are 4 errors reported by Checkstyle 6.17 with /builds/delphi/delphi/modules/delphi-main/src/test/google_checks.xml ruleset.
[ERROR] src/main/java/com/egnyte/delphi/cloudapp/common/command/ListPlugableCommandFactory.java:[26] (javadoc) SummaryJavadoc: First sentence of Javadoc is incomplete (period is missing) or not present.
[ERROR] src/main/java/com/egnyte/delphi/cloudapp/common/command/MapPlugableCommandFactory.java:[28] (javadoc) SummaryJavadoc: First sentence of Javadoc is incomplete (period is missing) or not present.
[ERROR] src/main/java/com/egnyte/delphi/cloudapp/egnyte/priv/PrivateApiEgnyteDomainModule.java:[21] (sizes) LineLength: Line is longer than 120 characters (found 124).
[ERROR] src/main/java/com/egnyte/delphi/cloudapp/egnyte/pub/PublicApiEgnyteDomainModule.java:[20] (sizes) LineLength: Line is longer than 120 characters (found 124).
```

> Have you tried to set up the Google style rules manually in SonarQube? I know it sounds like a pain, but there aren't that many rules and was just wondering if it's possible...
It should certainly be possible, but I spent the time instead setting up the flow described above :)

**Comment by timwebster9** [comment]
Ok - thanks again for the info...

**Comment by romani** [comment]
Hi all, I am going to release 3.1 version ..
there were done numerous updates to fix configurations and allow to use all Check and all their properties.

config should be ok to import by - https://github.com/checkstyle/sonar-checkstyle/wiki/How-to-import-existing-checkstyle-config-to-sonar after release.

I am closing this issue for now.

**Comment by romani** [comment]
there are still problem but different ....
Will be addressed in #34
